packages/frontend/core/src/hooks/affine/use-highlight.ts

```diff
@@ -0,0 +1,60 @@
+import { useMemo } from 'react';
+
+export function highlightTextFragments(text: string, query: string) {
+  const highlights = [];
+
+  const lowerCaseText = text.toLowerCase();
+  let startIndex = lowerCaseText.indexOf(query);
+
+  if (startIndex !== -1) {
+    // Full match found, highlight it
+    if (startIndex > 0) {
+      highlights.push({
+        text: text.substring(0, startIndex),
+        highlight: false,
+      });
+    }
+
+    highlights.push({
+      text: text.substring(startIndex, startIndex + query.length),
+      highlight: true,
+    });
+
+    if (startIndex + query.length < text.length) {
+      highlights.push({
+        text: text.substring(startIndex + query.length),
+        highlight: false,
+      });
+    }
+  } else {
+    // No full match found, so we proceed to partial matching
+    startIndex = 0;
+    for (const char of query) {
+      const pos = text.toLowerCase().indexOf(char, startIndex);
+      if (pos !== -1) {
+        if (pos > startIndex) {
+          highlights.push({
+            text: text.substring(startIndex, pos),
+            highlight: false,
+          });
+        }
+        highlights.push({
+          text: text.substring(pos, pos + 1),
+          highlight: true,
+        });
+        startIndex = pos + 1;
+      }
+    }
+    if (startIndex < text.length) {
+      highlights.push({
+        text: text.substring(startIndex),
+        highlight: false,
+      });
+    }
+  }
+  return highlights;
+}
+
+export function useHighlight(text: string, query: string) {
+  return useMemo(() => highlightTextFragments(text, query), [text, query]);
```

Comment:
if use iterator, this code cloud be `useMemo(() => Array.from(highlightTextFragments(text, query)), [text, query]);`